components/script/layout_dom/element.rs

```diff
@@ -457,6 +509,24 @@ impl<'dom> style::dom::TElement for ServoLayoutElement<'dom> {
             servo_layout_node.as_element().unwrap()
         }
     }
+
+    fn slotted_nodes(&self) -> &[Self::ConcreteNode] {
+        let Some(slot_element) = self.element.unsafe_get().downcast::<HTMLSlotElement>() else {
+            return &[];
+        };
+        let assigned_nodes = slot_element.assigned_nodes();
+
+        // SAFETY:
+        // Self::ConcreteNode (aka ServoLayoutNode) and Slottable are guaranteed to have the same
+        // layout and alignment as ptr::NonNull<T>. Lifetimes are not an issue because the
+        // slottables are being kept alive by the slot element.
+        unsafe {
+            slice::from_raw_parts(
+                assigned_nodes.as_ptr() as *const Self::ConcreteNode,
+                assigned_nodes.len(),
+            )
+        }
```

Comment:
This is the "interesting part" of this PR: Stylo wants the nodes stored inside a slot element as a borrowed slice. This requires some unsafe magic, because no one owns a list of `ServoLayoutNode`'s that could be borrowed here. However, since all `ServoLayoutNode`, `Dom` and `Slottable` all store a `NonNull<T>` (and have `#[repr(transparent)]`, it is safe to transmute here.

From my understanding it is okay to return a reference to DOM contents here for stylo to use.

For reference, gecko's implementation [^1] also contains significant amounts of `unsafe`.

[^1]: https://searchfox.org/mozilla-central/source/servo/components/style/gecko/wrapper.rs#1142